woocommerce/woocommerce-gutenberg-products-block
# Experimental newsletter subscription checkbox block POC and Store API supporting changes

This PR introduces some changes to the Store API to facilitate nested fields, an experimental newsletter subscription block, and additional hooks. This is part of #4448

**`useCheckoutExtensionData` hook**

This hook allows custom, namespaced data to be stored to the checkout-state context provider. Any values stored here are then passed to the Store API under the `extensions` top-level parameter. This has been implemented in the _newsletter_ block in this PR. It will need exporting externally for extensions to consume it.

_Testing:_

- Insert Newsletter block on checkout i2, save
- Go to frontend
- Check the box and submit order
- Check network inspector and confirm the newsletter API key was set on the request

**Store API changes**

In reviewing the Store API I noticed that nested parameters were never validated, so any given `arg_options` are ignored. WP core does not have a solution for this so I introduced a recursive validation function. Test this by editing the newsletter API registration in the main blocks plugin file in this PR:

```php
					'schema_callback' => function() {
						return array(
							'newsletter' => array(
								'description' => __( 'Subscribe to newsletter opt-in.', 'woo-gutenberg-products-block' ),
								'type'        => 'boolean',
								'context'     => array(),
								'arg_options' => array(
									'validate_callback' => function( $value ) {
										return new \WP_Error( 'api-error', ( $value ? 'true' : 'false' ) . ' was posted to the newsletter validation callback' );
									},
								),
							),
						);
					},
```

**Newsletter Block**

I've added a test newsletter block available within the contact section in order to test the above changes. This needs extracting into an extension to ensure all functionality is available externally. This is a follow-up item.
